Function definition can be assigned to variable.

nmdis1999 opened the pull request:
Major changes:
1.`vm.rs` has:
         a)  seperate `OpCode::InlineFunc` for defining anonymous functions.
         b) `Call` match considers 3 cases: finding target in `function`, `locals` and `label`. 
         c)  extended `StoreVar` match arm to store `String` as well. Later I'll add support for storing functions in variable too. 
               but I feel that should be seperate PR.
2). `compiler.rs` has:
        a) `CallTarget` enum to seperate choices between `functions`, anon function assigned to variables, and `label` such as  
              `print`.
        b) `FuncDef` now handles cases for `InlineFunc`.
3) `ukiyo.rs` has:
        a) grammar addition for annon function.
4) Added test file `inline_func.ukiyo`

ptersilie commented on src/lib/compiler.rs:
```diff
@@ -34,14 +42,15 @@ impl fmt::Display for OpCode {
             OpCode::StoreVar(i) => write!(f, "StoreVar({})", i),
             OpCode::LoadVar(i) => write!(f, "LoadVar({})", i),
-            OpCode::Call(s) => write!(f, "Call({})", s),
+            OpCode::Call(s) => todo!(),
```
Did you mean to change this to a todo?

ptersilie commented on src/lib/ukiyo.y:
```diff
@@ -33,13 +33,19 @@ param_list -> Result<Vec<Expr>, ()>:
          param_list { $1 }
        ;
-
+              
```
Random whitespace.

ptersilie commented on src/lib/vm.rs:
```diff
@@ -72,12 +74,30 @@ fn vm(
+                if let Some(val) = stack.pop() {
+                    let cval = val.to_owned();
+                    match val {
+                        Types::Int(x) => {
```
I think you might be able to collapse `Int` and `String`, since you're not using the contained value:
```
Types::Int(_) | Types::String(_) => { ... }
```
I haven't tested this though, so I might be wrong.

ptersilie commented on src/lib/vm.rs:
```diff
@@ -86,37 +106,50 @@ fn vm(
-                            Types::NoneType => output.push_str("None"),
+            OpCode::Call(ct) => {
+                match ct {
+                    crate::compiler::CallTarget::Func(label) => {
```
If you update the `use` at the top of the file by adding `CallTarget` you can shorten this a bit.

nmdis1999 replied:
Addressed here: c283d6acbae0c7319e587e2aa556806a2e12d39d

ptersilie commented on src/lib/vm.rs:
```diff
@@ -86,37 +106,50 @@ fn vm(
                         }
-                        stack.push(val_copy);
                     }
+                    crate::compiler::CallTarget::Label(label) => {
```
`Label` is a slightly confusing name for this. Maybe `Builtins` will make clearer these aren't user functions?

nmdis1999 replied:
Agreed, will combine your syntax & naming suggetion in one commit :)

ptersilie commented on src/lib/vm.rs:
```diff
@@ -217,6 +259,7 @@ pub fn run(ast: Ast, lexer: &dyn NonStreamingLexer<DefaultLexeme<u32>, u32>) ->
     let prog = compiler(ast, lexer);
     let mut locals = Vec::new();
     let mut functions: Vec<Function> = Vec::new();
+    dbg!(&prog);
```
Left a debug print.

ptersilie commented on tests/files/inline_func.ukiyo:
```diff
@@ -0,0 +1,13 @@
+    let fn = func(a, b) {
+        return a + b;
+    };
+    let x = fn();
```
I'm not sure how this test is working, since you're not passing any arguments to this function. I would expect this to throw an error. Which actually might be a good test to add, e.g. that calling a function with wrong arguments (too few, too many) throws an error.

nmdis1999 replied:
Addressed here: 6f0cfbfd64d0a8878927f991451cc3200ad201e6

ptersilie commented:
Just a couple of comments.

nmdis1999 commented:
Addressed review comments: formatting related commit: c283d6acbae0c7319e587e2aa556806a2e12d39d
and 6f0cfbfd64d0a8878927f991451cc3200ad201e6 for verifying arguments are passed right.

ptersilie commented on src/lib/vm.rs:
```diff
@@ -116,10 +120,14 @@ fn vm(
+                                println!("Error: Incorrect number of arguments. Expected {} arguments, but {} were provided.", func.args.len(), args_len);
+                                panic!();
+                            }
                             let result = vm(func_prog, locals, functions)?;
```
Ah, and now I understand why your previous test still worked. You're passing the parents locals into the function call here. This works purely by accident. If you add another test this will quickly break, e.g.
```
func foo() {
    return func (a, b) {
        return a + b;
    }
}

func main() {
    let f = foo();
    f();
}
```

nmdis1999 replied:
Addressed here: af8b860e411566e9f74f2ddc57ac340038ae8ece

ptersilie commented on tests/files/inline_func.ukiyo:
```diff
@@ -1,13 +1,13 @@
-//     
+
 func foo() {
     let a = 1; let b = 2;
```
If you remove this line, the test will also break, I think.

nmdis1999 replied:
It doesn't fail but I have restored this anyways.

ptersilie commented:
Another comment regarding the new test.

ptersilie commented on src/lib/compiler.rs:
```diff
@@ -232,9 +232,9 @@ fn compiler_expr(
+                bc[offset] = OpCode::InlineFunc(args, func_body, new_locals);
             } else {
-                bc[offset] = OpCode::DefineFunc(func_name.clone(), args, func_body);
+                bc[offset] = OpCode::DefineFunc(func_name.clone(), args, func_body, new_locals);
```
Why are we both passing `new_locals` into the function while at the same time doing `args.extend(new_locals)` above? Do we really need both?

ptersilie commented on src/lib/vm.rs:
```diff
@@ -102,34 +99,45 @@ fn vm(
+                        if let Some(index) =
+                            functions.iter().position(|f| f.name == Some(label.clone()))
+                        {
+                            let mut func = functions[index].clone();
```
Why do we need to clone `func` here?

ptersilie commented on src/lib/vm.rs:
```diff
@@ -102,34 +99,45 @@ fn vm(
                             }
                             let result = vm(func.prog.clone(), &mut localsnew, functions)?;
                             stack.extend(result);
+                            func.locals = Some(localsnew);
```
We do we assign to `func.locals` here. Since `func` is a clone and immediately falls out or scope here, I would expect this to do nothing.

ptersilie commented on src/lib/vm.rs:
```diff
@@ -227,23 +235,31 @@ fn vm(
-            OpCode::InlineFunc(args, func_prog) => {
+            OpCode::InlineFunc(args, func_prog, locals) => {
+                let mut func_locals = Vec::new();
+                for local in locals.iter() {
```
This is a great candidate for `map`, e.g.
```rust
let func_locals = locals.iter().map(|x| Types::String(x.clone())).collect();
```

ptersilie commented on src/lib/vm.rs:
```diff
@@ -227,23 +235,31 @@ fn vm(
                     args: args.to_vec(),
                     prog: func_prog.to_vec(),
-                    locals: None,
+                    locals: Some(func_locals),
```
Now I'm confused. We never passed `locals` into a function before? How did this work before this change?

ptersilie replied:
I just checked and I can't see `locals` being read anywhere. What am I missing?

ptersilie commented:
Added some comments.

ptersilie commented:
> Addressed comments @ptersilie in [cf3fd58](https://github.com/softdevteam/ukiyo_lang/commit/cf3fd588594deb2cf8f03e89d1225516c03b9d46) commit :) Apologies for delays.

Great commit. 1 addition, 13 deletions. That's what I like to see. And we got rid of a `clone` too. :+1: 

Please squash.

nmdis1999 commented:
Done 00a4b1c18933de884d462613ea2e29297e36ba8a, thank you!

ptersilie commented:
bors r+

ptersilie commented:
bors r-

ptersilie commented:
Just saw there's still a conflict. We might have to rebase first.

ptersilie commented:
> Just saw there's still a conflict. We might have to rebase first.

Actually, make this is a merge commit first. So just pull in master, resolve the conflict and push it. After I checked we can rebase.

nmdis1999 commented:
I think the conflicts are resolved, can you lmk if this looks good?

ptersilie commented:
Since you force pushed I can't really see what changed during the merge. Typically there's no need to force push when doing a merge commit. Anyway, please squash and then we see if bors is happy.

ptersilie commented:
bors r+
